design-documents/graph-ql/coverage/deprecate-createEmptyCart-add-createCart.md

```diff
@@ -0,0 +1,40 @@
+# Add `createCart`, deprecate `createEmptyCart`
+
+## What
+
+- Mark `createEmptyCart` as deprecated
+- Add `createCart` as a replacement
+
+## Why
+
+This PR came from feedback provided by [@sirugh](https://github.com/sirugh) from the [`pwa-studio`](https://github.com/magento/pwa-studio) team.
+
+When a cart is needed for a user, the UI sends a `mutation` for `createEmptyCart`. Immediately after that value is obtained, the UI then makes a `query` for the details of the cart that was just created.
+
+It's not ideal that a common flow like this requires 2 round-trips, one after the other.
+
+## Proposed Changes
+
+```diff
+type Mutation {
++    createCart(
++        input: CreateCartInput
++    ): CreateCartOutput @doc(description: "Create a new shopping cart")
```

Comment:
The current `createEmptyCart` mutation has some magic to it. If the request is authenticated with a bearer token the mutation returns the user's existing cart id otherwise it returns a new cart's id. 

I see in the new response type the `cart` is returned, which is ideal, but would it also still return a user's existing cart?

Reply:
> If the request is authenticated with a bearer token the mutation returns the user's existing cart id otherwise it returns a new cart's id.

Does it return the existing cart ID even in the case where that cart _is not_ empty? If so that makes the `createEmptyCart` name a bit counter-intuitive.

Reply:
> Does it return the existing cart ID even in the case where that cart is not empty

Checked, it does. 

- [`createCartForCustomer->execute`](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/QuoteGraphQl/Model/Resolver/CreateEmptyCart.php#L70) in resolver
- [`createCustomerCart`](https://github.com/magento/magento2/blob/c28466437c5b6867bd1d517c184e51d690ebfc46/app/code/Magento/Quote/Model/QuoteManagement.php#L269) in `createEmptyCartForCustomer`
- [`quoteRepository->getActiveForCustomer`](https://github.com/magento/magento2/blob/c28466437c5b6867bd1d517c184e51d690ebfc46/app/code/Magento/Quote/Model/QuoteManagement.php#L351) in `createCustomerCart`

That's confusing! I think the name is more confusing than the API, because you wouldn't want to flush a customer's existing cart, but a customer can only have a single cart.

I wonder what a better name is here? A name that makes it clear that this returns an existing thing, or falls back to creating a new thing.

Reply:
Yes, it returns the user's existing cart id in both cases. And yes, I agree. That's why we alias over it in [our code](https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/queries/createCart.graphql).

Reply:
> I wonder what a better name is here? A name that makes it clear that this returns an existing thing, or falls back to creating a new thing.

In PWA Studio a user can only have one cart visible and active. We have no concept of fetching data for a "stashed" cart. So I can't see why you ever need the concept of `new` cart. I would think that simply querying for `cart` would suffice regardless. You're either getting a new cart or an active one if authed.

Reply:
> So I can't see why you ever need the concept of new cart

Y'all need it for guest users, right? You don't need the concept of create for logged-in because there's a 1:1 relationship between customer/cart, so the server can hide creation from the client. But for a guest you'd need to ask to create _something_ to track cart items.

---

A bit confused about the "stashed cart" here. I'm talking about these 2 use-cases specifically that the `createEmptyCart` API handles today:

1. Getting existing cart (or quietly creating new one) if logged-in
2. Creating a guest cart

You mentioned the behavior works well, but we agree the name `createEmptyCart` is off. Just trying to identify a name that better describes that functionality

Reply:
Sorry - shouldn't have even mentioned "stashed cart". I don't even know if Magento has the idea of a customer having multiple carts at once.

`getCart` seems reasonable, but the name sounds more like a query.

Reply:
Went and looked at venia a bit more, and now I'm questioning if `createEmptyCart` API handling logged-in users at all.

@sirugh what's the situation where we would want to use `createEmptyCart` for a logged-in user? It seems like, at the time venia sends this mutation, we already should know if the user is a guest or not, right?

I think these are the cases we need to cover, but _please_ correct me if I'm wrong:

Use Case | Covered by
-- | --
Fetch cart for logged-in user | Query.customerCart
Fetch cart for guest user | Query.cart
Get a cart id to use for guest user | Mutation.createEmptyCart

If the above is correct, is seems like `createEmptyCart` should be something like `createGuestCart`, and you can skip that call entirely when a user is logged-in. Thoughts?

Reply:
> getCart seems reasonable, but the name sounds more like a query.

Yeah, I keep wanting to say it, but then stop myself each time because it doesn't make it clear that it's a side-effecty operation.

Hopefully my proposal above this comment improves things?

Reply:
> @sirugh what's the situation where we would want to use createEmptyCart for a logged-in user? It seems like, at the time venia sends this mutation, we already should know if the user is a guest or not, right?

I don't remember why we used the single mutation but it seemed to make the logic simpler.

> If the above is correct, is seems like createEmptyCart should be something like createGuestCart. Thoughts?

Sure that sounds fine.

Reply:
> I don't remember why we used the single mutation but it seemed to make the logic simpler.

Makes sense. I guess we ended up just moving the logic into an odd place: a `createEmptyCart` query that can return a non-empty cart.

I pushed up some changes to the proposal based on this thread. Can you give it another read through?